examples/puppeteer/oauth2-test.js

```diff
@@ -0,0 +1,117 @@
+ *   "password": "your-password"
+ * }
+ */
+const config = require('../config.keys.json');
+
+async function main() {
+  const oAuth2Client = await getAuthenticatedClient();
+  const url = 'https://www.googleapis.com/plus/v1/people?query=pizza';
+  const res = await oAuth2Client.request({ url });
+  console.log(res.data);
+}
+
+function getAuthenticatedClient() {
+  let browser = null;
+  return new Promise((resolve, reject) => {
+    // create an oAuth client to authorize the API call.  Secrets are kept in a `keys.json` file,
+    // which should be downloaded from the Google Developers Console.
+    const oAuth2Client = new OAuth2Client(
+      keys.web.client_id,
+      keys.web.client_secret,
+      keys.web.redirect_uris[0]
+    );
+
+    // Generate the url that will be used for the consent dialog.
+    const authorizeUrl = oAuth2Client.generateAuthUrl({
+      access_type: 'offline',
+      scope: 'https://www.googleapis.com/auth/plus.me',
+      prompt: 'consent'
+    });
+
+    // Open an http server to accept the oauth callback. In this simple example, the
+    // only request to our webserver is to /oauth2callback?code=<code>
+    const server = http
+      .createServer(async (req, res) => {
+        if (req.url.indexOf('/oauth2callback') > -1) {
+          // acquire the code from the querystring, and close the web server.
+          const qs = querystring.parse(url.parse(req.url).query);
+          console.log(`Code is ${qs.code}`);
+          res.end('Authentication successful! Please return to the console.');
+          server.close();
+
+          // Now that we have the code, use that to acquire tokens.
+          const r = await oAuth2Client.getToken(qs.code);
+          // Make sure to set the credentials on the OAuth2 client.
+          oAuth2Client.setCredentials(r.tokens);
+          console.info('Tokens acquired.');
+          await browser.close();
+          resolve(oAuth2Client);
+        }
+      })
+      .listen(3000, async () => {
+        // open the browser to the authorize url to start the workflow
+        browser = await doLogin(authorizeUrl);
+      });
+  });
+}
+
+/**
+ * Use the puppeteer library to sign into Google
+ */
+async function doLogin(authorizeUrl) {
+  const browser = await puppeteer.launch({ headless: false });
+  const page = await browser.newPage();
+  await page.goto(authorizeUrl);
+  await page.mainFrame().waitForSelector('#identifierId');
+  console.log('typing email...');
+  await page.type('#identifierId', config.username);
+  await page.mainFrame().waitForSelector('#identifierNext');
+  console.log('clicking next button...');
+  await page.click('#identifierNext');
+  console.log('waiting for password field...');
+  await page
+    .mainFrame()
+    .waitForSelector('#password input[type="password"]', { visible: true });
+  console.log('typing password...');
+  await page.type('#password input[type="password"]', config.password, {
+    delay: 100
```

Comment:
Why is a delay needed here?

Reply:
UI automation jank.  Puppeteer is fairly new, and it was hard to get the timing of all this stuff to work right.